src/midifunctor.cpp

```diff
@@ -799,9 +801,15 @@ FunctorCode GenerateMIDIFunctor::VisitStaffDef(const StaffDef *staffDef)
 FunctorCode GenerateMIDIFunctor::VisitSyl(const Syl *syl)
 {
     const double startTime = m_totalTime + m_lastNote->GetScoreTimeOnset();
-    const std::string sylText = UTF32to8(syl->GetText());
 
-    m_midiFile->addLyric(m_midiTrack, startTime * m_midiFile->getTPQ(), sylText);
+    m_currentNoteLyrics.append(!m_currentNoteLyrics.empty(), ' ');
+    m_currentNoteLyrics.append(UTF32to8(syl->GetText()));
+
+    const Object *parent = syl->GetParent();
+    assert(parent);
+    if (parent->GetNext(syl, SYL) == NULL) {
+        m_midiFile->addLyric(m_midiTrack, startTime * m_midiFile->getTPQ(), m_currentNoteLyrics);
```

Comment:
Just move this to a new `GenerateMIDIFunctor::VisitVerseEnd` and remove the lookup here.